Generic/JumbledIncrementer: fix non-problematic bug and improve test coverage

rodrigoprimo opened the pull request:
## Description

This PR adds a few commits related to the Generic.CodeAnalysis.JumbledIncrementer sniff:

- Fix a non-problematic bug that would happen when handling an inner for loop without a closing parenthesis. In those cases, `$end` would be set to `null` ([src/Standards/Generic/Sniffs/CodeAnalysis/JumbledIncrementerSniff.php#L116](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/84acf4e56f110db8e75cb9a575c5727df637643c/src/Standards/Generic/Sniffs/CodeAnalysis/JumbledIncrementerSniff.php#L116)) instead of set to the index of the closing parenthesis token. This problem did not cause any issues as `$next <= $end` when `$end` is `null` is always false, and the for loop that looks for incrementers ends early.
- Improve test coverage
- ~~Improve the name of a variable~~

A few things that I noticed while working on this PR that I would like to confirm if they are indeed issues (probably in separate PRs):

- The sniff documentation and its name only mention incrementers, but the sniff also works for decrementers. Should we update the documentation? I guess the name is a bit more complex to change, so I'm inclined to think we should not change it.
- The sniff [assumes that any variable in the third part of the for loop condition is an incrementer or decrementer](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/a42b5eacd7c9648a652218b0c09f9fdb24138485/src/Standards/Generic/Sniffs/CodeAnalysis/JumbledIncrementerSniff.php#L124-L126) which is not true. The code below triggers the sniff and I'm not sure if it should:

```php
$label = 'label';

for ($i = 1; $i <= 10; $i++, print $label) {
    for ($j = 0; $j < 5; $j++, print $label);
}
```

## Suggested changelog entry

Generic.CodeAnalysis.JumbledIncrementer: fixed a non-problematic bug when the inner for loop is missing the closing parenthesis

## Related issues/external references

Part of #146


## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix _(non-breaking change which fixes an issue)_
- [ ] New feature _(non-breaking change which adds functionality)_
- [ ] Breaking change _(fix or feature that would cause existing functionality to change)_
    - [ ] This change is only breaking for integrators, not for external standards or end-users.
- [ ] Documentation improvement


## PR checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have checked there is no other PR open for the same change.
- [x] I have read the [Contribution Guidelines](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/master/.github/CONTRIBUTING.md).
- [x] I grant the project the right to include and distribute the code under the BSD-3-Clause license (and I have the right to grant these rights).
- [x] I have added tests to cover my changes.
- [x] I have verified that the code complies with the projects coding standards.
- [ ] [Required for new sniffs] I have added XML documentation for the sniff.

<!--
============================================================================================
Please make sure your pull request passes all continuous integration checks!

PRs which are failing their CI checks will likely be ignored by the maintainers.

Small PRs using atomic, descriptive commits are hugely appreciated as it will make
reviewing your changes easier for the maintainers.
============================================================================================
-->

jrfnl commented on src/Standards/Generic/Sniffs/CodeAnalysis/JumbledIncrementerSniff.php:
```diff
@@ -85,11 +85,11 @@ public function process(File $phpcsFile, $stackPtr)
 
             $inner = $this->findIncrementers($tokens, $tokens[$start]);
-            $diff  = array_intersect($outer, $inner);
+            $commonIncrementers = array_intersect($outer, $inner);
```
While I agree that the name `$diff`is not the best name, `$commonIncrementers` isn't either as it contains the assumption that variables will always be incremented, while they could also be decremented, or even be changed via a function call or some other calculation.

Maybe it should be called `$commonVariables` or `$varIntersection` or something ? (I'm not sure what the best name is, but the current name change, IMO, is not an improvement).

rodrigoprimo replied:
As we talked about during our call, we won't be changing this variable name at this time and we will leave it for when the other problems you found are addressed in a separate issue.

rodrigoprimo replied:
I will go ahead, drop the commit that changed this variable name and force push the branch.

jrfnl commented on src/Standards/Generic/Tests/CodeAnalysis/JumbledIncrementerUnitTest.1.inc:
```diff
@@ -0,0 +1,83 @@
+    }
+}
+
+for (; $i < 10; $i++) {
```
Just a small note here: test 👍🏻 , but it might help to add a small comment above each test case saying what it is testing, so the test case on line 27 is the "okay" case for `for` loops with not all expressions set, while this test is the "flag me" case for `for` loops with not all expressions set.

Could possible also be done by renaming the `$i` variable to `$same` for this "flag me" code sample ? That should make it easier to spot.

jrfnl replied:
Oh and if you do change the variable name, I can imagine you may also want to do this for the pre-existing test cases.

Commit wise, would probably be good to do this in a separate commit before adding the extra test cases. (bug no biggie if you mix it in with this commit or do this in  a commit where you change it for the whole test case file in one go).

rodrigoprimo replied:
Done! As you suggested, I added a commit before adding the extra test cases, renaming the variables in the pre-existing tests, and then another commit, renaming the variables in the new tests (which should be combined with the commit that adds the new tests before merging).

jrfnl commented on src/Standards/Generic/Tests/CodeAnalysis/JumbledIncrementerUnitTest.1.inc:
```diff
@@ -0,0 +1,83 @@
+    endfor;
+endfor;
+
+// Sniff bails early when there is no incrementor in the outer for loop.
```
```suggestion
// Sniff bails early when there is no incrementor in the third expression of the outer for loop.
```

jrfnl replied:
Looks like this fix-up was undone by the variable renaming commit ?

rodrigoprimo replied:
Actually, I made this change on purpose but unfortunately, I failed to highlight that in the commit message. Sorry for the confusion.

When working on the variable renaming commit, I thought that the code comment would be more precise without referencing an incrementer, since the sniff is triggered for decrementer as well. However, I did not realize that I was changing a code comment that had been changed based on your suggestion in the previous commit.

That being said, since the PR is already merged, I think it is fine to keep the comment as is. It is just a detail.

jrfnl commented on src/Standards/Generic/Tests/CodeAnalysis/JumbledIncrementerUnitTest.3.inc:
```diff
@@ -0,0 +1,6 @@
+<?php
+
+// Intentional parse error (missing for body).
```
```suggestion
// Intentional parse error (missing for conditions).
```

jrfnl commented on src/Standards/Generic/Tests/CodeAnalysis/JumbledIncrementerUnitTest.1.inc:
```diff
@@ -0,0 +1,83 @@
+<?php
```
For the "Skip for loop without body" part (~line 66), you've now added a test with a parse error, which is good, but there is actually valid code which could also trigger that code. I think it would be good to add a test with valid code too.

```php
for ($i = 0; $i < 20; $i++);
```

Possibly also add a test where the outer loop has the scope opener/closer, but the inner loop does not ?

rodrigoprimo replied:
I added a new commit with the two tests.

jrfnl reviewed (commented):
@rodrigoprimo Thank you for this PR and my apologies for taking a while before reviewing it.

Generally speaking, all looking good! Thank you for making these changes.

There's a few small nitpicks, but nothing major.

While reviewing this PR, I noticed a few other things the sniff does not correctly take into account. I will open separate issues for those.

rodrigoprimo commented:
Thanks for the review, @jrfnl!

I believe I addressed the points that you raised. I had to force push commits to the branch as I dropped the commit that renamed a variable in the sniff code and added a new commit in between commits to rename variables for the pre-existing tests.

This PR is ready for a final check.

jrfnl commented:
As per the comment I left in my previous review, I've now opened two issues for potential follow-up improvements for the sniff: #440 and #441

I also realized that I hadn't answered the questions you included in the description yet.

> The sniff documentation and its name only mention incrementers, but the sniff also works for decrementers. Should we update the documentation? I guess the name is a bit more complex to change, so I'm inclined to think we should not change it.

Good catch and yes, please feel invited to improve the documentation (docblock + XML ?).

And as you already surmised, changing the sniff name would be a breaking change, so is not on the table.

> The sniff [assumes that any variable in the third part of the for loop condition is an incrementer or decrementer](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/a42b5eacd7c9648a652218b0c09f9fdb24138485/src/Standards/Generic/Sniffs/CodeAnalysis/JumbledIncrementerSniff.php#L124-L126) which is not true. The code below triggers the sniff and I'm not sure if it should:

I fully agree with you that this is a bug. Issue #441 contains a range of additional code samples running into the same issue. Please add your code sample in a comment to that ticket so it won't be overlooked by whomever starts working on that ticket.

jrfnl reviewed (approved):
@rodrigoprimo Thank you for working on this and making the requested updates.

I'll go ahead and reorganize the commits and rebase the PR before merging it.

I will fix up that comment change which got undone while doing that.
